pkg/lib/queueinformer/queueinformer_operator.go

```diff
@@ -1,273 +1,300 @@
+	o.logger.Info("starting workers...")
+	for _, queueInformer := range o.queueInformers {
+		go o.worker(ctx, queueInformer)
+		go o.worker(ctx, queueInformer)
 	}
-	return loop.syncHandler(obj)
+
+	close(o.ready)
+	<-ctx.Done()
 }
 
-// This provides the same function as above, but for queues that are not auto-fed by informers.
-// indexerWorker runs a worker thread that just dequeues items, processes them, and marks them done.
+// worker runs a worker thread that just dequeues items, processes them, and marks them done.
 // It enforces that the syncHandler is never invoked concurrently with the same key.
-func (o *Operator) indexerWorker(loop *QueueIndexer) {
-	for o.processNextIndexerWorkItem(loop) {
+func (o *operator) worker(ctx context.Context, loop *QueueInformer) {
+	for o.processNextWorkItem(ctx, loop) {
 	}
 }
 
-func (o *Operator) processNextIndexerWorkItem(loop *QueueIndexer) bool {
+func (o *operator) processNextWorkItem(ctx context.Context, loop *QueueInformer) bool {
 	queue := loop.queue
-	key, quit := queue.Get()
+	item, quit := queue.Get()
 
 	if quit {
 		return false
 	}
-	defer queue.Done(key)
+	defer queue.Done(item)
+
+	logger := o.logger.WithField("item", item)
+	logger.WithField("queue-length", queue.Len()).Trace("popped queue")
+
+	event, ok := item.(kubestate.ResourceEvent)
+	if !ok || event.Type() != kubestate.ResourceDeleted {
+		// Get the key
+		key, keyable := loop.key(item)
+		if !keyable {
+			logger.WithField("item", item).Warn("could not form key")
+			queue.Forget(item)
+			return true
+		}
+
+		logger = logger.WithField("cache-key", key)
+
+		// Get the current cached version of the resource
+		resource, exists, err := loop.indexer.GetByKey(key)
+		if err != nil {
+			logger.WithError(err).Error("cache get failed")
+			// queue.Forget(item)
+			return true
+		}
+		if !exists {
+			logger.WithField("existing-cache-keys", loop.indexer.ListKeys()).Debug("cache get failed, key not in cache")
+			queue.Forget(item)
+			return true
+		}
 
-	// requeue five times on error
-	if err := o.syncIndexer(loop, key.(string)); err != nil && queue.NumRequeues(key.(string)) < 5 {
-		o.Log.Infof("retrying %s", key)
-		utilruntime.HandleError(errors.Wrap(err, fmt.Sprintf("Sync %q failed", key)))
-		queue.AddRateLimited(key)
+		if !ok {
+			event = kubestate.NewResourceEvent(kubestate.ResourceUpdated, resource)
+		} else {
+			event = kubestate.NewResourceEvent(event.Type(), resource)
+		}
+	}
+
+	// Sync and requeue on error (throw out failed deletion syncs)
+	err := loop.Sync(ctx, event)
+	if requeues := queue.NumRequeues(item); err != nil && requeues < 8 && event.Type() != kubestate.ResourceDeleted {
+		logger.WithField("requeues", requeues).Trace("requeuing with rate limiting")
+		utilruntime.HandleError(errors.Wrap(err, fmt.Sprintf("sync %q failed", item)))
+		queue.AddRateLimited(item)
 		return true
 	}
-	queue.Forget(key)
-	if err := loop.HandleMetrics(); err != nil {
-		o.Log.Error(err)
+	queue.Forget(item)
+
+	select {
```

Comment:
What is this select for?

Reply:
This was in the original `QueueInformer` code. It looks like it makes the channel send non-blocking, so the sync can exit without a receiver on the other side (or it's a full buffered channel).